Hi @bcoca ,

I tested this PR with two scenarios:
- The user launches `ansible all -m command -a "sleep 30"` and then, hit `Ctrl-C` before completion
- The user launches `ansible all -m command -a "sleep 30"` and then, send a `TERM` signal to the `ansible` process.

Here are my results:

#### When the user hit `Ctrl-C`:

##### Version 2.0.0.2

###### Output:

```
$ ansible -i /tmp/inventory.ini all -m command -a "sleep 30"
^C [ERROR]: User interrupted execution
```

###### Effect:

All the `ansible` processes are immediately killed.

##### With this PR `die_now`

###### Output:

```
$ PYTHONPATH=./lib/ python2 bin/ansible -i /tmp/inventory.ini all -m command -a "sleep 30"
^CProcess WorkerProcess-4:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/process/worker.py", line 115, in run
    self._shared_loader_obj,
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/task_executor.py", line 421, in _execute
    result = self._handler.run(task_vars=variables)
  File "/home/lenaic/doc/prog/ansible/lib/ansible/plugins/action/normal.py", line 33, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars))
  File "/home/lenaic/doc/prog/ansible/lib/ansible/plugins/action/__init__.py", line 463, in _execute_module
    res = self._low_level_execute_command(cmd, sudoable=sudoable, in_data=in_data)
  File "/home/lenaic/doc/prog/ansible/lib/ansible/plugins/action/__init__.py", line 524, in _low_level_execute_command
    rc, stdout, stderr = self._connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)
  File "/home/lenaic/doc/prog/ansible/lib/ansible/plugins/connection/local.py", line 118, in exec_command
    stdout, stderr = p.communicate()
  File "/usr/lib/python2.7/subprocess.py", line 799, in communicate
    return self._communicate(input)
  File "/usr/lib/python2.7/subprocess.py", line 1409, in _communicate
    stdout, stderr = self._communicate_with_poll(input)
  File "/usr/lib/python2.7/subprocess.py", line 1463, in _communicate_with_poll
    ready = poller.poll()
KeyboardInterrupt
Process WorkerProcess-3:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/process/worker.py", line 115, in run
    self._shared_loader_obj,
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/task_executor.py", line 119, in run
Process WorkerProcess-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/process/worker.py", line 115, in run
    self._shared_loader_obj,
ERROR! Unexpected Exception: can only test a child process
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/task_executor.py", line 119, in run
to see the full traceback, use -vvv
    res = self._execute()
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/task_executor.py", line 421, in _execute
    result = self._handler.run(task_vars=variables)
  File "/home/lenaic/doc/prog/ansible/lib/ansible/plugins/action/normal.py", line 33, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars))
  File "/home/lenaic/doc/prog/ansible/lib/ansible/plugins/action/__init__.py", line 463, in _execute_module
    res = self._low_level_execute_command(cmd, sudoable=sudoable, in_data=in_data)
  File "/home/lenaic/doc/prog/ansible/lib/ansible/plugins/action/__init__.py", line 524, in _low_level_execute_command
    rc, stdout, stderr = self._connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)
  File "/home/lenaic/doc/prog/ansible/lib/ansible/plugins/connection/local.py", line 118, in exec_command
    stdout, stderr = p.communicate()
  File "/usr/lib/python2.7/subprocess.py", line 799, in communicate
 [ERROR]: User interrupted execution

    return self._communicate(input)
  File "/usr/lib/python2.7/subprocess.py", line 1409, in _communicate
    stdout, stderr = self._communicate_with_poll(input)
  File "/usr/lib/python2.7/subprocess.py", line 1463, in _communicate_with_poll
    ready = poller.poll()
KeyboardInterrupt
```

###### Effect:

All the `ansible` processes are immediately killed.

#### When a `TERM` signal is sent to the father `ansible` process. (The one whose PID is the PPID of the others)

##### Version 2.0.0.2

###### Output:

```
$ ansible -i /tmp/inventory.ini all -m command -a "sleep 30"
Complété
```

###### Effect:

Only the killed process disappears immediately.
All the remote processes still continue to process.
When a remote process eventually finishes, the local `ansible` process corresponding completes.
At the end, when all the remote processes have finished, a local `ansible` process remains forever.

##### With this PR `die_now`

###### Output:

```
$ PYTHONPATH=./lib/ python2 bin/ansible -i /tmp/inventory.ini all -m command -a "sleep 30"
Traceback (most recent call last):
Interrupt detected, shutting down gracefully
  File "/home/lenaic/doc/prog/ansible/lib/ansible/executor/process/result.py", line 104, in run
Interrupt detected, shutting down gracefully
Interrupt detected, shutting down gracefully
Interrupt detected, shutting down gracefully
    time.sleep(0.0001)
  File "/home/lenaic/doc/prog/ansible/lib/ansible/cli/adhoc.py", line 94, in _terminate
    raise SystemExit("Interrupt detected, shutting down gracefully")
SystemExit: Interrupt detected, shutting down gracefully
```

###### Effect:

All the local `ansible` processes are immediately killed.
All the remote processes still continue to process.
When all the remote processes eventually finish, everything is clean and no stale process remains anywhere.

### Conclusion

With this PR, in both use cases (User hitting `Ctrl-C` and `TERM` signal sent to `ansible`), all the processes spawned by `ansible` eventually finished.

However, in the case of the `TERM` signal sent to `ansible`, the remote processes were not killed and they eventually finished because it was a `sleep`, but it they were stuck, they would have remained stuck forever.

When the user hits `Ctrl-C`, there is now a stack trace which is dumped. Since this use case is supposed to be gracefully handled, it would nicer to not display it.

--- a/lib/ansible/cli/adhoc.py
+++ b/lib/ansible/cli/adhoc.py
@@ -21,6 +21,7 @@ __metaclass__ = type
 ########################################################
 
 import os
+import signal
 
 from ansible import constants as C
 from ansible.cli import CLI
@@ -87,6 +88,10 @@ class AdHocCLI(CLI):
             gather_facts = 'no',
             tasks = [ dict(action=dict(module=self.options.module_name, args=parse_kv(self.options.module_args)), async=async, poll=poll) ]
         )
+
+    def _terminate(self, signum=None, framenum=None):
+        if signum is not None:
+            raise SystemExit("Interrupt detected, shutting down gracefully")
 
     def run(self):
         ''' use Runner lib to do SSH things '''
@@ -170,6 +175,9 @@ class AdHocCLI(CLI):
         # now create a task queue manager to execute the play
         self._tqm = None
         try:
+            # Manage user interruptions
+            signal.signal(signal.SIGTERM, self._terminate)
+
             self._tqm = TaskQueueManager(
                     inventory=inventory,
                     variable_manager=variable_manager,
@@ -180,6 +188,7 @@ class AdHocCLI(CLI):
                     run_additional_callbacks=C.DEFAULT_LOAD_CALLBACK_PLUGINS,
                     run_tree=run_tree,
                 )
+
             result = self._tqm.run(play)
         finally:
             if self._tqm:

--- a/lib/ansible/executor/playbook_executor.py
+++ b/lib/ansible/executor/playbook_executor.py
@@ -69,7 +69,7 @@ class PlaybookExecutor:
         may limit the runs to serialized groups, etc.
         '''
 
-        signal.signal(signal.SIGINT, self._cleanup)
+        signal.signal(signal.SIGTERM, self._terminate)
 
         result = 0
         entrylist = []
@@ -199,7 +199,7 @@ class PlaybookExecutor:
 
         finally:
             if self._tqm is not None:
-                self._cleanup()
+                self._tqm.cleanup()
 
         if self._options.syntax:
             display.display("No issues encountered")
@@ -207,8 +207,9 @@ class PlaybookExecutor:
 
         return result
 
-    def _cleanup(self, signum=None, framenum=None):
-        return self._tqm.cleanup()
+    def _terminate(self, signum=None, framenum=None):
+        display.debug(framenum)
+        raise SystemExit("Terminating run due to external signal")
 
     def _get_serialized_batches(self, play):
         '''

--- a/lib/ansible/executor/process/result.py
+++ b/lib/ansible/executor/process/result.py
@@ -163,7 +163,7 @@ class ResultProcess(multiprocessing.Process):
 
             except queue.Empty:
                 pass
-            except (KeyboardInterrupt, IOError, EOFError):
+            except (KeyboardInterrupt, SystemExit, IOError, EOFError):
                 break
             except:
                 # TODO: we should probably send a proper callback here instead of

--- a/lib/ansible/executor/process/worker.py
+++ b/lib/ansible/executor/process/worker.py
@@ -132,7 +132,7 @@ class WorkerProcess(multiprocessing.Process):
             self._rslt_q.put(task_result, block=False)
 
         except Exception as e:
-            if not isinstance(e, (IOError, EOFError, KeyboardInterrupt)) or isinstance(e, TemplateNotFound):
+            if not isinstance(e, (IOError, EOFError, KeyboardInterrupt, SystemExit)) or isinstance(e, TemplateNotFound):
                 try:
                     self._host.vars = dict()
                     self._host.groups = []
@@ -140,7 +140,7 @@ class WorkerProcess(multiprocessing.Process):
                     self._rslt_q.put(task_result, block=False)
                 except:
                     debug(u"WORKER EXCEPTION: %s" % to_unicode(e))
-                    debug(u"WORKER EXCEPTION: %s" % to_unicode(traceback.format_exc()))
+                    debug(u"WORKER TRACEBACK: %s" % to_unicode(traceback.format_exc()))
 
         debug("WORKER PROCESS EXITING")
 

--- a/lib/ansible/executor/task_queue_manager.py
+++ b/lib/ansible/executor/task_queue_manager.py
@@ -253,7 +253,10 @@ class TaskQueueManager:
                 rslt_q.close()
                 main_q.close()
                 if worker_prc and worker_prc.is_alive():
-                    worker_prc.terminate()
+                    try:
+                        worker_prc.terminate()
+                    except AttributeError:
+                        pass
 
     def clear_failed_hosts(self):
         self._failed_hosts = dict()

